File: majiang.c
#include <stdio.h>
int main(void){
  char player[4][16];
  int money[4] = {0};
  int loss[4] = {0};
  int win[4] = {0};
  int i, tmp;
  int lastWinner = -1;
  int maxComboWinner = -1;
  int tmpCombo = 0;
  int maxCombo = -1;
  int gameCount = 0;
  int lastLoser = -1;
  int maxBadComboLoser = -1;
  int tmpBadCombo = 0;
  int maxBadCombo = -1;
  for(i = 0; i < 4; i++)
    scanf("%s", player[i]);
  i = 0;
  while(scanf("%d", &tmp) != EOF){
    if(i == 0)
      gameCount++;
    money[i] += tmp;
    if(tmp < 0){
      loss[i]++;
      if(i == lastLoser){
        tmpBadCombo++;
        if(tmpBadCombo > maxBadCombo){
          maxBadCombo = tmpBadCombo;
          maxBadComboLoser = lastLoser;
        }
      }
      else{
        lastLoser = i;
        tmpBadCombo = 1;
      }
    }
    else if(tmp > 0){
      win[i]++;
      if(i == lastWinner){
        tmpCombo++;
        if(tmpCombo > maxCombo){
          maxCombo = tmpCombo;
          maxComboWinner = lastWinner;
        }
      }
      else{
        lastWinner = i;
        tmpCombo = 1;
      }
    }
    i++;
    i %= 4;
  }
  printf("        Total: %d games\n", gameCount);
  for(i = 0; i < 4; i++)
    printf("%7s ", player[i]);
  printf("\n");

  for(i = 0; i < 4; i++)
    printf("%7d ", money[i]);
  printf("\n");
  printf("        Losses\n");
  for(i = 0; i < 4; i++)
    printf("%7d ", loss[i]);
  printf("\n");
  printf("        Wins\n");
  for(i = 0; i < 4; i++)
    printf("%7d ", win[i]);
  printf("\n");
  printf("        Combo\n");
  printf("     %d-winning, made by %s\n", maxCombo, player[maxComboWinner]);
  printf("     %d-losing, made by %s\n", maxBadCombo, player[maxBadComboLoser]);
}

File: majiang.py
# encoding: utf-8
import sys
money = [0, 0, 0, 0]
loss = [0, 0, 0, 0]
win = [0, 0, 0, 0]
lineCount = 0
lastLoser = lastWinner = totallyLoser = totallyWinner = maxLoseCombo = maxWinCombo = -1
tmpLoseCombo = tmpWinCombo = 0
for line in sys.stdin:
    if lineCount == 0:  #get players' names
        players = line
    else:
        i = 0
        for tmp in line.replace('\n', '').split(' '):
            money[i] += int(tmp)
            if int(tmp) < 0:
                loss[i] -= 1
                if lastLoser == i:
                    tmpLoseCombo += 1
                    if tmpLoseCombo > maxLoseCombo:
                        maxLoseCombo = tmpLoseCombo
                        totallyLoser = lastLoser
                else:
                    lastLoser = i
                    tmpLoseCombo = 1
            elif int(tmp) > 0:
                win[i] += 1
                if lastWinner == i:
                    tmpWinCombo += 1
                    if tmpWinCombo > maxWinCombo:
                        maxWinCombo = tmpWinCombo
                        totallyWinner = lastWinner
                else:
                    lastWinner = i
                    tmpWinCombo = 1
            i += 1
            i %= 4
    lineCount += 1

gameCount = lineCount - 1
print "         Total:", gameCount, "games"

for name in (players.replace('\n', '')).split(' '):
    print '{:>7}'.format(name),
print
for record in money:
    print '{0:7d}'.format(record),
print
print "              Losses"
for record in loss:
    print '{0:7d}'.format(record),
print
print "              Wins"
for record in win:
    print '{0:7d}'.format(record),
print
print "              Combo"
print "    ", maxLoseCombo, "losses, made by", (players.replace('\n', '')).split(' ')[totallyLoser]
print "    ", maxWinCombo, "wins, made by", (players.replace('\n', '')).split(' ')[totallyWinner]
